### PR TITLE
use dns srv format for sni-dnat clusters

### PR DESCRIPTION
**Pull request by rshriram**
Istio generated SNI is using DNS srv format. So, SNI-DNAT clusters
need to follow same format so that traffic can be routed to appropriate clusters.

Signed-off-by: Shriram Rajagopalan <shriramr@vmware.com>

### CODE REVIEW

File: pilot/pkg/model/service.go [hunk 1]
```diff
@@ -822,7 +829,17 @@ func IsValidSubsetKey(s string) bool {
...
 	if len(parts) < 4 {
-		return
+		// This could be the DNS srv form of the cluster that uses outbound_.port_.subset_.hostname
+		// Since we dont want every callsite to implement the logic to differentiate between the two forms
+		// we add an alternate parser here.
+		if !strings.HasPrefix(s, fmt.Sprintf("%s_", TrafficDirectionOutbound)) &&
+			!strings.HasPrefix(s, fmt.Sprintf("%s_", TrafficDirectionInbound)) {
+			return
+		}
+		parts = strings.SplitN(s, ".", 4)
```
Comment (baodongli) [inline pilot/pkg/model/service.go]: It's safer to do Split(s, ".") and check the number of parts. The above SplitN() call doesn't preclude more then 4 parts.

File: pilot/pkg/model/service.go [hunk 2]
```diff
@@ -822,7 +829,17 @@ func IsValidSubsetKey(s string) bool {
...
+		parts[0] = strings.TrimSuffix(parts[0], "_")
+		parts[1] = strings.TrimSuffix(parts[1], "_")
+		parts[2] = strings.TrimSuffix(parts[2], "_")
```
Comment (baodongli) [inline pilot/pkg/model/service.go]: Would the logic be clearer: 
  if (strings.HasPrefix(s, fmt.Sprintf("%s_", TrafficDirectionOutbound)) ||
 			strings.HasPrefix(s, fmt.Sprintf("%s_", TrafficDirectionInbound) {
          parts = strings.Split(s, ".")
          ......
}  else {
          parts := strings.Split(s, "|")
}

File: pilot/pkg/model/service.go (hunk 1 above)
Comment (rshriram) [reply pilot/pkg/model/service.go]: we only need 4 parts. Also, this is not coming from end user. This is generated by cluster.go. So we know that the cluster name has 4 parts definitely

File: pilot/pkg/model/service.go [hunk 3]
```diff
@@ -813,20 +813,45 @@ func BuildSubsetKey(direction TrafficDirection, subsetName string, hostname Host
 	return fmt.Sprintf("%s|%d|%s|%s", direction, port, subsetName, hostname)
 }
 
+// BuildDNSSrvSubsetKey generates a unique string referencing service instances for a given service name, a subset and a port.
+// The proxy queries Pilot with this key to obtain the list of instances in a subset.
+// This is used only for the SNI-DNAT router. Do not use for other purposes.
+func BuildDNSSrvSubsetKey(direction TrafficDirection, subsetName string, hostname Hostname, port int) string {
+	return fmt.Sprintf("%s_.%d_.%s_.%s", direction, port, subsetName, hostname)
```
Comment (costinm) [inline pilot/pkg/model/service.go]: Do we really need direction ? Are there 2 directions even possible for DNAT ? 
Also, can we use the more common _ in front of the name (like in dnssd) ?

Having a domain component start with a number is a bit risky - I think it was not allowed in the 
original RFC, was relaxed later, but may still cause some confusion. Having _ in front would avoid this.

Comment (costinm) [reply pilot/pkg/model/service.go]: I see you use direction_ to detect it's of this time - we could still use a prefix, like:
  _istio._8080._v1.products.ns1.svc.cluster.local

Comment (rshriram) [reply pilot/pkg/model/service.go]: This name maps with the cluster names. Really don’t want to add another special string there. 

The underscore in the end is dns srv format isn’t it? GRPC and others use the srv format right?

I don’t mind changing it but will have to change the SNI format as well.

**Pull request merged**